### ISSUE TITLE
scaled_dot_product_attention: different output between the reference implementation and the one from torch.nn.functional

### ISSUE THREAD

**Issue by Andrei-Aksionov** [pytorch/pytorch#119188]
### 🐛 Describe the bug

Hello there 👋 

From my understanding, the code from the [docs](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html) should produce numerically equivalent output to what `torch.nn.functional.scaled_dot_product_attention` produces. The difference should be only in the speed.
But, as one can see from the code snippet below, it's not True.

What's also interesting is that if a tensor is in `torch.float32` and is filled with random variables from `Uniform` distribution, then we have a full match. But with the same precision, but with random values from `Normal` distribution - there is a difference. 
If tensors are placed on a GPU and/or in a lower precision - the difference in outputs is much, much higher.

**Question**: is it expected or it's a bug?

```python
import math
from itertools import product

import torch
from torch.nn import functional as F

# Note: copy-paste from https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html
#   with two changes: `attn_bias` and `temp_mask` are placed on a the same device as `query`
# Efficient implementation equivalent to the following:
def scaled_dot_product_attention(
    query, key, value, attn_mask=None, dropout_p=0.0, is_causal=False, scale=None
) -> torch.Tensor:
    # Efficient implementation equivalent to the following:
    L, S = query.size(-2), key.size(-2)
    scale_factor = 1 / math.sqrt(query.size(-1)) if scale is None else scale
    attn_bias = torch.zeros(L, S, dtype=query.dtype, device=query.device)
    if is_causal:
        assert attn_mask is None
        temp_mask = torch.ones(L, S, dtype=torch.bool, device=query.device).tril(diagonal=0)
        attn_bias.masked_fill_(temp_mask.logical_not(), float("-inf"))
        attn_bias.to(query.dtype)

    if attn_mask is not None:
        if attn_mask.dtype == torch.bool:
            attn_bias.masked_fill_(attn_mask.logical_not(), float("-inf"))
        else:
            attn_bias += attn_mask
    attn_weight = query @ key.transpose(-2, -1) * scale_factor
    attn_weight += attn_bias
    attn_weight = torch.softmax(attn_weight, dim=-1)
    attn_weight = torch.dropout(attn_weight, dropout_p, train=True)
    return attn_weight @ value


torch.manual_seed(42)
B, nh, T, C = (1, 4, 29, 128)
def generate_qkv(dtype, device, rand_generator):
    return [rand_generator((B, nh, T, C), dtype=dtype, device=device) for _ in range(3)]

for dtype, rand_generator, device in product(
    (torch.float32, torch.float16, torch.bfloat16), (torch.rand, torch.randn), ("cpu", "cuda")
):
    if dtype is torch.float16 and device == "cpu":
        continue
    if dtype is torch.bfloat16 and device == "cuda" and not torch.cuda.is_bf16_supported():
        continue

    rand_name = "Uniform Distribution" if (rand_generator is torch.rand) else "Normal Distribution"
    print(f"{dtype} | {rand_name} | {device}")

    query, key, value = generate_qkv(dtype, device, rand_generator)

    reference_output = scaled_dot_product_attention(query, key, value, is_causal=True)
    functional_output = F.scaled_dot_product_attention(query, key, value, is_causal=True)
    print("IsClose frac:", torch.isclose(reference_output, functional_output).sum() / reference_output.nelement())
    print("-" * 100)
```

```text
torch.float32 | Uniform Distribution | cpu
IsClose frac: tensor(1.)
----------------------------------------------------------------------------------------------------
torch.float32 | Uniform Distribution | cuda
IsClose frac: tensor(1., device='cuda:0')
----------------------------------------------------------------------------------------------------
torch.float32 | Normal Distribution | cpu
IsClose frac: tensor(0.9993)
----------------------------------------------------------------------------------------------------
torch.float32 | Normal Distribution | cuda
IsClose frac: tensor(0.9979, device='cuda:0')
----------------------------------------------------------------------------------------------------
torch.float16 | Uniform Distribution | cuda
IsClose frac: tensor(0.8140, device='cuda:0')
----------------------------------------------------------------------------------------------------
torch.float16 | Normal Distribution | cuda
IsClose frac: tensor(0.4525, device='cuda:0')
----------------------------------------------------------------------------------------------------
torch.bfloat16 | Uniform Distribution | cpu
IsClose frac: tensor(0.8060)
----------------------------------------------------------------------------------------------------
torch.bfloat16 | Normal Distribution | cpu
IsClose frac: tensor(0.5158)
----------------------------------------------------------------------------------------------------
```

----

Resources:
- [Colab Notebook](https://colab.research.google.com/drive/1_E4RrOUWOS--7X_mIGq9VMzW6X1lycgQ?usp=sharing) with the code above

### Versions

Collecting environment information...
PyTorch version: 2.1.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.27.9
Libc version: glibc-2.35

Python version: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.1.58+-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.2.140
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 535.104.05
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.6
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             2
On-line CPU(s) list:                0,1
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) CPU @ 2.00GHz
CPU family:                         6
Model:                              85
Thread(s) per core:                 2
Core(s) per socket:                 1
Socket(s):                          1
Stepping:                           3
BogoMIPS:                           4000.39
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat md_clear arch_capabilities
Hypervisor vendor:                  KVM
Virtualization type:                full
L1d cache:                          32 KiB (1 instance)
L1i cache:                          32 KiB (1 instance)
L2 cache:                           1 MiB (1 instance)
L3 cache:                           38.5 MiB (1 instance)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0,1
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Mitigation; PTE Inversion
Vulnerability Mds:                  Vulnerable; SMT Host state unknown
Vulnerability Meltdown:             Vulnerable
Vulnerability Mmio stale data:      Vulnerable
Vulnerability Retbleed:             Vulnerable
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Vulnerable
Vulnerability Spectre v1:           Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:           Vulnerable, IBPB: disabled, STIBP: disabled, PBRSB-eIBRS: Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Vulnerable

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] torch==2.1.0+cu121
[pip3] torchaudio==2.1.0+cu121
[pip3] torchdata==0.7.0
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.16.0
[pip3] torchvision==0.16.0+cu121
[pip3] triton==2.1.0
[conda] Could not collect

**Comment by Lifedecoder** [comment]
I find same problem, that results get from function called is different from reimplementing its code.
Wonder that have you solved it?

**Comment by drisspg** [comment]
This is expected, while the fused implementations in SDPA are semantically equivalent. There are two main sources for this:
- Floating point numerics: https://pytorch.org/docs/stable/notes/numerical_accuracy.html has a good description of this concept
- The fused attention algorithms use iterative softmax which can introduce some numerical differences between the unfused implementation.

**Comment by Lifedecoder** [comment]
Are the differences that large? I used F.scaled_dot_product_attention and the reimplementing one, but got totally different results in inference. Is this expected too?

**Comment by drisspg** [comment]
It is hard to say without knowing the exact amount of deviation you are saying

**Comment by Lifedecoder** [comment]
I got two results here, upper is using F.scaled_dot_product_attention:
https://github.com/pytorch/pytorch/assets/53321710/e7d68c90-4752-46e2-b02c-744e209a5132

this one uses reimplementing scaled_dot_product_attention:
https://github.com/pytorch/pytorch/assets/53321710/3952634e-2319-4c6f-ab25-f89cca24b868

Seems abnormal, big deviation

**Comment by Lifedecoder** [comment]
More concretely, i find the attention weight calculated by
`attn_weight = query @ key.transpose(-2, -1) * scale_factor`
is all 'nan'.

I use torch==2.0.1, and whether i use official implementation from torch 2.1 and 2.0, attention weights are all nan.

I adjust the calculation order into 
`attn_weight = query @ (key.transpose(-2, -1) * scale_factor)`, attention weights are not nan, close to results given out by F.scaled_dot_product_attention (almost the same value for most elements), but the ultimate results are still not right(all black as shown above).

**Comment by Lifedecoder** [comment]
the inf problem is caused by overflow... convert to torch.float64 and calculate on cpu could handle, but very slow.
